source/opt/loop_unswitch_pass.cpp

```diff
@@ -0,0 +1,1179 @@
+                        opt::TreeDFIterator<ir::Loop>())) {
+      if (IsLoopDead(old_to_new_bb_.at(sub_loop.GetHeaderBlock()->id()),
+                     old_to_new_bb_.at(sub_loop.GetLatchBlock()->id()),
+                     dead_blocks))
+        continue;
+      std::unique_ptr<ir::Loop> cloned = MakeUnique<ir::Loop>();
+      if (ir::Loop* parent = get_parent(sub_loop.GetParent()))
+        parent->AddNestedLoop(cloned.get());
+      PopulateLoopDesc(cloned.get(), &sub_loop, dead_blocks,
+                       unreachable_merges);
+    }
+
+    std::for_each(
+        loop_mapping.begin(), loop_mapping.end(),
+        [this](
+            std::unordered_map<ir::Loop*, ir::Loop*>::iterator::reference it) {
+          if (!it.second->GetParent())
+            loop_desc_.AddLoops(std::unique_ptr<ir::Loop>(it.second));
+        });
+  }
+
+  // Populates |new_loop| descriptor according to |old_loop|'s one.
+  void PopulateLoopDesc(ir::Loop* new_loop, ir::Loop* old_loop,
+                        const std::unordered_set<uint32_t>& dead_blocks,
+                        std::unordered_set<uint32_t>* unreachable_merges) {
+    for (uint32_t bb_id : old_loop->GetBlocks()) {
+      ir::BasicBlock* bb = old_to_new_bb_.at(bb_id);
+      if (!dead_blocks.count(bb->id())) {
+        new_loop->AddBasicBlock(bb);
+      }
+    }
+    new_loop->SetHeaderBlock(
+        old_to_new_bb_.at(old_loop->GetHeaderBlock()->id()));
+    if (old_loop->GetLatchBlock())
+      new_loop->SetLatchBlock(
+          old_to_new_bb_.at(old_loop->GetLatchBlock()->id()));
+    if (old_loop->GetMergeBlock()) {
+      ir::BasicBlock* bb = old_to_new_bb_.at(old_loop->GetMergeBlock()->id());
+      // The merge block might be unreachable, in which case it will be tagged
+      // as dead. We need it, so mark it as must be kept.
+      if (dead_blocks.count(bb->id())) {
+        unreachable_merges->insert(bb->id());
+      }
+      new_loop->SetMergeBlock(bb);
+    }
+    if (old_loop->GetPreHeaderBlock())
+      new_loop->SetPreHeaderBlock(
+          old_to_new_bb_.at(old_loop->GetPreHeaderBlock()->id()));
+  }
+
+  // Returns true if |condition| is dynamically uniform.
+  // Note: this is currently approximated as uniform.
+  bool IsDynamicallyUniform(ir::Instruction* condition,
+                            const ir::BasicBlock* entry,
+                            const DominatorTree& post_dom_tree) {
+    assert(post_dom_tree.IsPostDominator());
+
+    auto it = dynamically_uniform_.find(condition->result_id());
+
+    if (it != dynamically_uniform_.end()) return it->second;
+
+    ir::BasicBlock* parent = context_->get_instr_block(condition);
+    if (!parent) {
+      return dynamically_uniform_[condition->result_id()] = true;
+    }
+    if (!post_dom_tree.Dominates(parent->id(), entry->id())) {
+      return dynamically_uniform_[condition->result_id()] = false;
+    }
+    bool uniform = condition->WhileEachInId(
```

Comment:
I think this will return true for a load from a "private" variable, but they are not uniform.  Two different threads might have different values in those locations.

I think the correct check it that the instruction must be a combinator whose parameters are all dynamically-uniform. For example, I don't think the derivative instructions should be considered dynamically uniform or anything that consumes the implicit result.  Somebody who knows more would have to comment on that @dneto0 .  If it is a load, it must be from a memory location in the uniform or uniform constant storage class.

Also, anything that has the "uniform" decoration is uniform.

I see in the future another pass that decorates anything that can be proven to be uniform, then all that is needed is to check the decorations.  I don't expect you to write that.

Reply:
I integrated your comment in this commit https://github.com/KhronosGroup/SPIRV-Tools/pull/1292/commits/f73c150349020df5aa0ca216923b10ba30710082